phoenix-core/src/main/java/org/apache/phoenix/schema/MetaDataClient.java

```diff
@@ -4168,4 +4176,197 @@ public MutationState useSchema(UseSchemaStatement useSchemaStatement) throws SQL
+            revokePermLog.append(" for Schema: " + revokeStatement.getSchemaName());
+        } else if (revokeStatement.getTableName() != null) {
+            revokePermLog.append(" for Table: " + revokeStatement.getTableName());
+        }
+        logger.info(revokePermLog.toString());
+
+        HConnection hConnection = connection.getQueryServices().getAdmin().getConnection();
+
+        try {
+            if (revokeStatement.getSchemaName() != null) {
+                // SYSTEM.CATALOG doesn't have any entry for "default" HBase namespace, hence we will bypass the check
+                if(!revokeStatement.getSchemaName().equals(QueryConstants.HBASE_DEFAULT_SCHEMA_NAME)) {
+                    FromCompiler.getResolverForSchema(revokeStatement.getSchemaName(), connection);
+                }
+                revokePermissionsFromSchema(hConnection, revokeStatement);
+
+            } else if (revokeStatement.getTableName() != null) {
+                PTable inputTable = PhoenixRuntime.getTable(connection,
+                        SchemaUtil.normalizeFullTableName(revokeStatement.getTableName().toString()));
+                if (!(PTableType.TABLE.equals(inputTable.getType()) || PTableType.SYSTEM.equals(inputTable.getType()))) {
+                    throw new AccessDeniedException("Cannot REVOKE permissions from INDEX TABLES or VIEWS");
+                }
+                revokePermissionsFromTables(hConnection, revokeStatement, inputTable);
+
+            } else {
+                revokePermissionsFromUser(hConnection, revokeStatement);
+            }
+
+        } catch (SQLException e) {
+            // Bubble up the SQL Exception
+            throw e;
+        } catch (Throwable throwable) {
+            // Wrap around other exceptions to PhoenixIOException (Ex: org.apache.hadoop.hbase.security.AccessDeniedException)
+            throw ServerUtil.parseServerException(throwable);
+        }
+
+        return new MutationState(0, 0, connection);
+    }
+
+
+    private void revokePermissionsFromTables(HConnection hConnection, RevokeStatement revokeStatement, PTable inputTable) throws Throwable {
+
+        org.apache.hadoop.hbase.TableName tableName = SchemaUtil.getPhysicalTableName
+                (inputTable.getName().getBytes(), inputTable.isNamespaceMapped());
+
+        revokePermissionsFromTable(hConnection, revokeStatement, tableName);
+
+        for(PTable indexTable : inputTable.getIndexes()) {
+            // Local Indexes don't correspond to new physical table, they are just stored in separate CF of base table.
+            if(indexTable.getIndexType().equals(IndexType.LOCAL)) {
+                continue;
+            }
+            logger.info("Revoking perms from IndexTable: " + indexTable.getName() + " BaseTable: " + inputTable.getName());
+            if (inputTable.isNamespaceMapped() != indexTable.isNamespaceMapped()) {
+                throw new TablesNotInSyncException(inputTable.getTableName().getString(),
```

Comment:
Can this ever happen (base table is namespace mapped but index table is not)?

Reply:
I don't think it can happen. indexes should go in the same namespace in which data table is.

Reply:
Theoretically yes, However since the operation is not atomic, it may result into weird case where they are differing.

Reply:
This is more like sanity check again.